R/join_keys-utils.R

```diff
@@ -111,44 +111,31 @@ update_keys_given_parents <- function(x) {
 
   datanames <- names(jk)
   duplicate_pairs <- list()
```

Comment:
could we allocate the size here instead of increasing it loop something like this
```
len <- length(datanames)
  duplicate_pairs <- vector("list", length = len * (len - 1) / 2)
  d_ix <- 1
```

and instead of 
`duplicate_pairs <- append(duplicate_pairs, paste(d1, d2))`
do this
```
duplicate_pairs[[d_ix]] <-  paste(d1, d2)
d_ix <- d_ix+1
```

Reply:
This function would benefit from some refactoring, but I don't think we get any performance benefits with the expected size of `join_keys`.

Given no performance boost _(assuming benchmarks below are realistic)_, I find it easier to read the code as is. I'm open to discuss it though :+1: 

The individual operations are more efficient as `f{1,2}_manual()` shows, however, that boost is completely lost when accounting for the initialization of the vector.

At least not until we get into real big vectors (see second benchmark), and still it needs to go up to absurd sizes.

``` r
ntimes <- 100L

f1 <- local({
  l <- NULL
  \() {
    if (is.null(l)) {
      print("initializing f1")
      l <<- list()
    }
    l <<- append(l, paste("ad", "bc"))
    invisible(l) # returning this or NULL does not affect benchmark
  }
})

f2 <- local({
  l <- NULL
  d_ix <- 1
  \() {
    if (is.null(l)) {
      print("initializing f2")
      l <<- vector("list", length = ntimes)
    }
    l[[d_ix]] <- paste("ad", "bc")
    d_ix <<- d_ix + 1
    invisible(l) # returning this or NULL does not affect benchmark
  }
})

f1_manual_l <- rep(paste("ad", "bc"), ntimes - 1) |> as.list()
f2_manual_l <- vector("list", length = ntimes)
d_ix <- length(f2_manual_l)

f1_manual <- \() invisible(append(f1_manual_l, paste("ad", "bc")))
f2_manual <- \() {
  f2_manual_l[[d_ix]] <- paste("ad", "bc")
  d_ix <- d_ix + 1
  invisible(f2_manual_l)
}

microbenchmark::microbenchmark(f1(), f2(), f1_manual(), f2_manual(), times = ntimes)
#> [1] "initializing f2"
#> [1] "initializing f1"
#> Unit: microseconds
#>         expr   min     lq     mean median     uq      max neval
#>         f1() 2.067 2.2405  2.58464 2.3640 2.4955   15.255   100
#>         f2() 2.181 2.2600  3.22357 2.2985 2.3720   87.602   100
#>  f1_manual() 2.285 2.3765 91.59518 2.4225 2.5110 8911.203   100
#>  f2_manual() 1.723 1.7850 22.44896 1.8250 1.8725 2055.476   100
```

``` r
ntimes <- 10000L

# ...

microbenchmark::microbenchmark(f1(), f2(), f1_manual(), f2_manual(), times = ntimes)
#> [1] "initializing f1"
#> [1] "initializing f2"
#> Unit: microseconds
#>         expr    min      lq     mean  median      uq      max neval
#>         f1()  2.186 16.0270 31.30275 29.4425 43.3985 1861.159 10000
#>         f2() 42.127 44.5890 52.01254 45.8570 50.1740 4813.170 10000
#>  f1_manual() 46.798 48.8735 56.40130 50.1150 55.0195 8137.664 10000
#>  f2_manual() 41.856 44.0300 51.30092 45.2985 49.3840 3963.022 10000
```

Reply:
> This function would benefit from some refactoring, but I don't think we get any performance benefits with the expected size of join_keys.

Yeah, I did not consider the size, so I don't think the list of join_keys will be very long to make any significant difference.

Reply:
That variable was not actually being used :facepalm: 

Regardless, went a bit deeper and tested a real benchmark actually filling the list at each run (and without cold starts)

I'm so sure anymore about the benefit (even for small sizes)

``` r
b <- \(max_size = 10, ntimes = 100L) {
  f1 <- local({
    l <- NULL
    l_call <- \(.x) {
      if (is.null(l)) l <<- list()
      l <<- append(l, paste("ad", "bc"))
      TRUE
    }
    \() vapply(seq_len(max_size), l_call, logical(1))
  })
  
  f2 <- local({
    l <- NULL
    d_ix <- 1
    l_call <- \(.x) {
      if (is.null(l)) l <<- vector("list", length = max_size)
      l[[d_ix]] <<- paste("ad", "bc")
      d_ix <<- d_ix + 1
      TRUE
    }
    \() vapply(seq_len(max_size), l_call, logical(1))
  })
  
  microbenchmark::microbenchmark(
    f1(), 
    f2(), 
    times = ntimes, 
    setup = {
      list()
      vector("list", length = max_size)
    }
  )
}

range <- 1:100

res <- Reduce(
  \(x, y) {
    res <- summary(b(y))
    dplyr::bind_rows(x, data.frame(max_size = y, f1 = res$mean[[1]], f2 = res$mean[[2]]))
  },
  x = range,
  init = data.frame(max_size = numeric(0L), f1 = numeric(0L), f2 = numeric(0L))
)

library(ggplot2)
res |> 
  dplyr::mutate(ratio = f1 / f2) |> 
  dplyr::rename(append = f1, assign = f2) |> 
  tidyr::pivot_longer(cols = c(append, assign)) |> 
  dplyr::mutate(
    ratio = dplyr::if_else(name == "append", as.character(round(ratio, digits = 2)), NA_character_),
    ratio = dplyr::if_else(max_size %% 5 == 0, ratio, NA_character_)
  ) |> 
  
  ggplot(mapping = aes(x = max_size, y = value, color = name, label = ratio)) + 
    geom_point() +
    ggrepel::geom_label_repel(min.segment.length = 0) + 
    scale_y_continuous(trans = "log10", name = "Mean in microseconds (log10 scale)") +
    labs(subtitle = "label is ratio between append / assign") +
    theme_minimal()
#> Warning: Removed 180 rows containing missing values (`geom_label_repel()`).
```

![](https://i.imgur.com/dpOEKYT.png)<!-- -->

<sup>Created on 2023-12-20 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>